src/day8/re_frame/test.cljc

```
@@ -1,66 +1,242 @@
+  [& body]
+  `(let [restore-fn# (make-re-frame-restore-fn)]
+     (try
+       ~@body
+       (finally
+         (restore-fn#)))))
+
+
+;;;;
+;;;; Async tests
+;;;;
+
+(def ^:dynamic *test-timeout* 5000)
+
+(def ^{:dynamic true, :private true} *test-context*
+  "`*test-context*` is used to communicate internal details of the test between
+  `run-test-async*` and `wait-for*`. It is dynamically bound so that it doesn't
+  need to appear as a lexical argument to a `wait-for` block, since we don't
+  want it to be visible when you're writing tests.  But care must be taken to
+  pass it around lexically across callbacks, since ClojureScript doesn't have
+  `bound-fn`."
+  nil)
+
+
+(defn run-test-async* [f]
+  (let [test-context {:wait-for-depth     0
+                      :max-wait-for-depth (atom 0)
+                      :now-waiting-for    (atom nil)}]
+    #?(:clj  (let [done-promise (promise)
+                   executor     (Executors/newSingleThreadExecutor)
+                   fail-ex      (atom nil)
+                   test-context (assoc test-context :done #(deliver done-promise ::done))]
+               (with-redefs [rf-int/executor executor]
+                 ;; Execute the test code itself on the same thread as the
+                 ;; re-frame event handlers run, so that we accurately
+                 ;; simulate the single-threaded JS environment and also so
+                 ;; that we don't have to worry about making the JVM
+                 ;; implementation of the re-frame EventQueue thread-safe.
+                 (rf-int/next-tick #(try
+                                      (binding [*test-context* test-context]
+                                        (f))
+                                      (catch Throwable t
+                                        (reset! fail-ex t)))))
+               (let [result (deref done-promise *test-timeout* ::timeout)]
+                 (doto executor
+                   (.shutdownNow)
```

Should we throw an error if there are outstanding tasks to run on the queue that haven't been started yet?